### PR TITLE
lexer: Tokenize expression using a scanner

### PR DESCRIPTION
**Pull request by qinqon**
For nmpolicy the parser will use a lexer to tokenize the expression into
tokens, it scans the runes using a `io.RunScanner`.

### CODE REVIEW

**Comment by qinqon** [comment]
@AlonaKaplan cannot add you as reviewer I supose is because the invitation is pending, although this is not finished yet.

**Comment by qinqon** [comment]
I am refactoring this to move the scanner part to its own package and create a new packge for error and source.

**Comment by AlonaKaplan** [comment]
> For nmpolicy the parser will use a lexer to tokenize the expression into tokens, it scans the runes using the reader from `source.Reader` at #3.

I think the PR message should b updated since the reference PR is closed.

File: nmpolicy/internal/lexer/lexer.go [hunk 1]
```diff
@@ -19,24 +19,88 @@
...
+		}
+
+		if l.isEOF() || l.isSpace() {
+			// If it's EOF or space we have finish here
+			return token, nil
+		} else if l.isDigit() {
+			token.Literal += string(l.scn.Rune())
+		} else {
+			// nmpolicy supports only simple numbers with just digist
+			return nil, fmt.Errorf("invalid number format (%s is not a digit)", string(l.scn.Rune()))
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer.go:122]: Does it mean we require a space before a pipe/merge? If so, we need to update some of our design doc examples.

File: nmpolicy/internal/lexer/lexer_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	type expected struct {
+		tokens []lexer.Token
+		err    string
+	}
+
+	var tests = []struct {
+		expression string
+		expected   expected
+	}{
+		{"    ", expected{tokens: []lexer.Token{
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer_test.go:35]: Please add tests with `+` and `|` as part of the expression. e.g "1|5", "12 5+7"

File: nmpolicy/internal/lexer/rune.go [hunk 3]
```diff
@@ -31,3 +32,7 @@ func (l *lexer) isSpace() bool {
 func (l *lexer) isEOF() bool {
 	return l.scn.Rune() == scanner.EOF
 }
+
+func (l *lexer) isString() bool {
+	return strings.ContainsRune(`"'`, l.scn.Rune())
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/rune.go:37]: As far as I know, yamls support strings with no quotes at all. Do we want to block it for nmpolicy?

File: .golangci.yml [hunk 4]
```diff
@@ -116,3 +116,8 @@ linters:
   # - testpackage
   # - revive
   # - wsl
+issues:
```
Comment (AlonaKaplan) [inline .golangci.yml:119]: How is this related to the commit?

File: nmpolicy/internal/lexer/lexer.go [hunk 5]
```diff
@@ -100,6 +104,11 @@ func (l *lexer) lexNumber() (*Token, error) {
 		if l.isEOF() || l.isSpace() {
 			// If it's EOF or space we have finish here
 			return token, nil
+		} else if l.isDot() {
+			if err := l.scn.Prev(); err != nil {
+				return nil, fmt.Errorf("failed lexing identity: %w", err)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer.go:115]: s/failed lexing number

File: nmpolicy/internal/lexer/rune.go [hunk 6]
```diff
@@ -40,3 +40,15 @@ func (l *lexer) isString() bool {
...
+
+func (l *lexer) isDot() bool {
+	return l.scn.Rune() == '.'
+}
+
+func (l *lexer) isEqual() bool {
+	return l.scn.Rune() == '='
+}
+
+func (l *lexer) isVerticalTwoDots() bool {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/rune.go:52]: `isColon`

File: nmpolicy/internal/lexer/lexer.go [hunk 7]
```diff
@@ -86,6 +86,10 @@ func (l *lexer) lexCurrentRune() (*Token, error) {
 		return l.lexString()
 	} else if l.isLetter() {
 		return l.lexIdentity()
+	} else if l.isDot() {
+		return &Token{l.scn.Position(), DOT, string(l.scn.Rune())}, nil
+	} else if l.isVerticalTwoDots() {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer.go:91]: Please update the commit message, the commit also adds support to replace.

File: nmpolicy/internal/lexer/lexer.go [hunk 8]
```diff
@@ -146,7 +148,7 @@ func (l *lexer) lexIdentity() (*Token, error) {
 
 		if l.isEOF() || l.isSpace() {
 			return token, nil
-		} else if l.isDot() || l.isEqual() || l.isVerticalTwoDots() {
+		} else if l.isDot() || l.isEqual() || l.isVerticalTwoDots() || l.isPlus() {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer.go]: Same should be added for `lexNumber` isn't it?
EDIT: since currently we support only merging of captures, we actually don't need it here atm.

File: nmpolicy/internal/lexer/lexer.go [hunk 9]
```diff
@@ -148,7 +150,7 @@ func (l *lexer) lexIdentity() (*Token, error) {
 
 		if l.isEOF() || l.isSpace() {
 			return token, nil
-		} else if l.isDot() || l.isEqual() || l.isVerticalTwoDots() || l.isPlus() {
+		} else if l.isDot() || l.isEqual() || l.isVerticalTwoDots() || l.isPlus() || l.isPipe() {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer.go:155]: Same should be added for lexNumber isn't it?

File: .golangci.yml [hunk 10]
```diff
@@ -121,3 +121,4 @@ issues:
     - path: _test.go
       linters:
         - funlen
+        - dupl
```
Comment (AlonaKaplan) [inline .golangci.yml:124]: How is this related to the commit?

File: nmpolicy/internal/lexer/lexer.go (hunk 1 above)
Comment (qinqon) [reply nmpolicy/internal/lexer/lexer.go:119]: At this commit we just parse numbers, what do you mean with pipe/merge ?

File: nmpolicy/internal/lexer/lexer_test.go (hunk 2 above)
Comment (qinqon) [reply nmpolicy/internal/lexer/lexer_test.go:37]: This is the commit for numbers.

File: nmpolicy/internal/lexer/rune.go (hunk 3 above)
Comment (qinqon) [reply nmpolicy/internal/lexer/rune.go:39]: This is not yaml, this is the capture expressions.

File: nmpolicy/internal/lexer/lexer.go (hunk 8 above)
Comment (qinqon) [reply nmpolicy/internal/lexer/lexer.go]: From token perspective if we find the plus rune the identity has finish.

File: nmpolicy/internal/lexer/lexer.go (hunk 9 above)
Comment (qinqon) [reply nmpolicy/internal/lexer/lexer.go:152]: As said, if it's not a number we have finish.

File: .golangci.yml (hunk 10 above)
Comment (qinqon) [reply .golangci.yml]: Removing

File: nmpolicy/internal/lexer/lexer_test.go (hunk 2 above)
Comment (AlonaKaplan) [reply nmpolicy/internal/lexer/lexer_test.go]: I know, I want the tests to tests "1|5" and "12 5+7" fails tokenization. If you want to add it to a further commit it's fine.

File: nmpolicy/internal/lexer/lexer.go [hunk 11]
```diff
@@ -145,7 +147,7 @@ func (l *lexer) lexIdentity() (*Token, error) {
 
 		if l.isEOF() || l.isSpace() {
 			return token, nil
-		} else if l.isDot() || l.isEqual() || l.isColon() || l.isPlus() {
+		} else if l.isDot() || l.isEqual() || l.isColon() || l.isPlus() || l.isPipe() {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/lexer/lexer.go:152]: Regarding, I meant that if you have `some.path==5|another.path:="whatever" - iiuc,, tokenizing the number 5 will fail.
So `|| l.isPipe()` should be added to `lexNumber` (line 105) as well.

File: nmpolicy/internal/lexer/lexer.go (hunk 8 above)
Comment (qinqon) [reply nmpolicy/internal/lexer/lexer.go]: Fixed

**Review by AlonaKaplan — APPROVED** [review]
